File: Trees.py


class Node:
    def __init__(self, value):
        self.val = value
        self.left = None
        self.right = None
    

class BST:
    def __init__():
        self.root = None
    

#   Create an add(val) method on the BST object to add new value to the tree
    # def add(self, val):
    #     new_node = Node(val)
    #     if self.root == None:
    #         self.root = new_node
    #         return self
    #     runner = self.root
    #     while runner != None:
    #         if val <= runner.value:
    #             if runner.left == None:
    #                 runner.left = new_node
    #             runner = runner.left
    #         else:
    #             if runner.right == None:
    #                 runner.right = new_node
    #             runner = runner.right
    #     return self
    
    def add(self, val):
        if val == self.val:
            return
        if val < self.val:
            if self.left:
                self.left.add(val)
            else:
                self.left=BST(val)
        else:
            if self.right:
                self.right.add(val)
            else:
                self.right = BST(val)


#   Create a contains(val) method on BST that returns whether the tree contains a given value
    def contains(self, val):
        if val == self.val:
            return True
        if val < self.val:
            if self.left == None:
                return False
            return self.left.exists(val)

        if self.right == None:
            return False
        return self.right.exists(val)

#   Create a min() method on the BST class that returns the smallest value found in the BST
    def min(self):
        current = self
        while current.left != None:
            current = current.left
        return current.val

#   Create a max() BST method that returns the largest value contained in the binary search tree
    def max(self):
        current = self
        while current.right != None:
            current = current.right
        return current.val


    # def PrintTree(self):
    #     if self.left:
    #         self.left.PrintTree()
    #     print( self.data),
    #     if self.right:
    #         self.right.PrintTree()



# bst=BST(5)
bst=BST(5)
bst.add(2)
bst.add(6)
bst.add(1)
bst.add(8)
print(bst.value)

File: trees.js
class Node {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}
class BST {
    constructor() {
        this.root = null;
    }


    // ADD
    add(val) {
        if(this.root) {
            var runner = this.root;
            while(runner) {
                if(val>runner.value) {
                    if(runner.right){
                        runner = runner.right;
                    } else {
                        runner.right = new Node(val);
                        return this
                    }
                } else {
                    if(runner.left){
                        runner = runner.left;
                    } else {
                        runner.left = new Node(val);
                        return this
                    }
                }
            }
        }
        this.root = new Node(val);
        return this
    }
    contains(val) {
        var runner = this.root;
        while (runner) {
            if (val== runner.val) {
                return true;
            }
            if (val < runner.val) {
                if (!runner.left) {
                    return false;
                }
                runner = runner.left;
            } else {
                if (!runner.right) {
                    return false;
                }
                runner = runner.right;
            }
        }
        return false;
    }
    min() {
        var runner = this.root;
        var min = this.root.val;
        while(runner.left) {
            if(runner.left.val < min) {
                min = runner.left.val;
            }
            runner = runner.left;
        }
        return min
    }
    max() {
        var runner = this.root;
        var max = this.root.val;
        while(runner.right) {
            if(runner.right.val > max) {
                max = runner.right.val;
            }
            runner = runner.right;
        }
        return max
    }
}
// creating a new BST
tree = new BST();

// add
tree.add(2).add(4).add(5).add(7).add(3).add(12).add(19).add(17).add(6).add(4)

// contains
console.log(tree.contains(17))
console.log(tree.contains('hello'))

// min
console.log(tree.min())
//max
console.log(tree.max())

console.log(tree);